server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -146,7 +255,133 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+                break;
+            default:
+                break;
+        }
+    }
+
+    private void diagnoseDeciders(
+        BiConsumer<UserAction.Definition, ShardRouting> diagnosisOutput,
+        ShardRouting shardRouting,
+        ClusterState state
+    ) {
+        LOGGER.trace("Diagnosing shard [{}]", shardRouting.shardId());
+        RoutingAllocation allocation = new RoutingAllocation(
+            allocationDeciders,
+            state,
+            ClusterInfo.EMPTY,
+            SnapshotShardSizeInfo.EMPTY,
+            System.nanoTime()
+        );
+        allocation.setDebugMode(RoutingAllocation.DebugMode.ON);
+        ShardAllocationDecision shardAllocationDecision = allocationService.explainShardAllocation(shardRouting, allocation);
+        AllocateUnassignedDecision allocateDecision = shardAllocationDecision.getAllocateDecision();
+        LOGGER.trace(
+            "[{}]: Obtained decision: [{}/{}]",
+            shardRouting.shardId(),
+            allocateDecision.isDecisionTaken(),
+            allocateDecision.getAllocationDecision()
+        );
+        if (allocateDecision.isDecisionTaken() && AllocationDecision.NO == allocateDecision.getAllocationDecision()) {
+            if (LOGGER.isTraceEnabled()) {
+                LOGGER.trace(
+                    "[{}]: Working with decisions: [{}]",
+                    shardRouting.shardId(),
+                    allocateDecision.getNodeDecisions()
+                        .stream()
+                        .map(
+                            n -> n.getCanAllocateDecision()
+                                .getDecisions()
+                                .stream()
+                                .map(d -> d.label() + ": " + d.type())
+                                .collect(Collectors.toList())
+                        )
+                        .collect(Collectors.toList())
+                );
+            }
+            List<NodeAllocationResult> nodeAllocationResults = allocateDecision.getNodeDecisions();
+            boolean diagnosisFound = checkIsAllocationDisabled(diagnosisOutput, shardRouting, nodeAllocationResults);
+            IndexMetadata index = state.metadata().index(shardRouting.index());
+            if (index != null) {
+                diagnosisFound |= checkDataTierRelatedIssues(diagnosisOutput, index, shardRouting, nodeAllocationResults);
+            }
+            if (diagnosisFound == false) {
+                diagnosisOutput.accept(ACTION_CHECK_ALLOCATION_EXPLAIN_API, shardRouting);
+            }
+        }
+    }
+
+    private static Predicate<NodeAllocationResult> nodeHasDeciderResult(String deciderName, Decision.Type outcome) {
+        return (nodeResult) -> nodeResult.getCanAllocateDecision()
+            .getDecisions()
+            .stream()
+            .anyMatch(decision -> deciderName.equals(decision.label()) && outcome == decision.type());
+    }
+
+    private boolean checkIsAllocationDisabled(
```

Comment:
I think these methods should return the definition/routing rather than take the consumer and do a builder-y type of buildup. They're much easier to think about because you can see exactly where the output is used rather than passing a consumer through that you have to go look up its actual function.

Reply:
very much agreed - refactoring is in [948b5e2](https://github.com/elastic/elasticsearch/pull/85071/commits/948b5e29b176071ffedb7613fa99563bb2d213e9), [ec52fca](https://github.com/elastic/elasticsearch/pull/85071/commits/ec52fca4b95cb98e7016272d465b42bf48cf489d)